### PR TITLE
Official type definitions

### PR DESCRIPTION
**Pull request by javiertury**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `next` branch for v4 (or to a previous version branch), _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing: https://github.com/vuelidate/vuelidate/blob/master/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

Implements #723

I believe the typings cover all of vuelidate functionality except `$autoDirty`.

### CODE REVIEW

**Comment by dobromir-hristov** [comment]
Hey this is awesome. I have some feedback that I will give tomorrow early morning.

It would be better if we rewrite Vuelidate in TS, but once we get to a more stable phase.

`$autoDirty` goes away, as it makes no sense to define it for each value. We will have top validation settings for such things, like lazyValidators and so on.

**Review by dobromir-hristov — COMMENTED** [review]
First of all, great job! I hope my JSDoc typehints were a bit of a help, even if they are a bit outdated here and there.

I have left a few minor comments. Frankly I cant review this as my TS skills are pretty basic as of now. Those types at the end gave me a migraine 😆

File: packages/vuelidate/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+ *            |            |
+ *           Validation Result
+ *
+ */
+
+export type ValidatorFn <T = unknown> = (value: T) => boolean;
+
+export interface ValidationRuleWithoutParams <T = unknown> {
+  $validator: ValidatorFn<T>
+  $message?: string
```
[the remark this answers is not in the document]
Comment (javiertury) [reply packages/vuelidate/index.d.ts]: I agree

**Comment by javiertury** [comment]
> I hope my JSDoc typehints were a bit of a help

They are very helpful to learn the structure of validation objects and map them to typescript definitions. Thank you :)

File: packages/vuelidate/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  readonly $error: boolean
+  readonly $errors: ErrorObject[]
+  readonly $invalid: boolean
+  readonly $anyDirty: boolean
+  readonly $pending: boolean
+
+  // const validationMethods
+  readonly $touch: () => void
+  readonly $reset: () => void
+  // readonly $validate?: () => Promise<boolean>
```
[the remark this answers is not in the document]
Comment (javiertury) [reply packages/vuelidate/index.d.ts]: Here I have a question, is `$getResultsForChild` recursive?

```
validation.$getResultsForChild('foo').$getResultsForChild('bar')...
```

[the remark this answers is not in the document]
Comment (dobromir-hristov) [reply packages/vuelidate/index.d.ts]: It should be. It returns the Entire validation state of the child you fetch.

**Comment by dobromir-hristov** [comment]
Btw I did a small refactor on most validations - #724 Turns out then they get transpiled and bundled, the output was a bit weird for the IDE to easily typehint fro the jsdoc. Anyway if you can double check your validation types with those, would be awesome!

**Comment by javiertury** [comment]
I've updated the definitions for rules, now most accept `Ref`s as parameters.

**Comment by dobromir-hristov** [comment]
Awesome, you think we can merge it in?

**Comment by javiertury** [comment]
I've built and polished them over the course of 2-3 months. I cannot guarantee them to be completely bug free, but I believe the definitions cover most of the functionality and are very precise.

**Pull request merged**